### PR TITLE
Fix elapsed time calculations

### PR DESCRIPTION
**Pull request by bogdanvlviv**
Hello,

I've found a few places in Rails code base where I think it makes sense to calculate elapsed time more precisely by using `Concurrent.monotonic_time`:

- `ActiveSupport::Cache::MemoryStore#prune`
- `ActiveRecord::ConnectionAdapters::ConnectionPool::Queue#wait_poll`
- `ActiveRecord::ConnectionAdapters::ConnectionPool#attempt_to_checkout_all_existing_connections`
- `ActiveRecord::ConnectionAdapters::Mysql2Adapter#explain`

See
https://docs.ruby-lang.org/en/2.5.0/Process.html#method-c-clock_gettime
https://blog.dnsimple.com/2018/03/elapsed-time-with-ruby-the-right-way

Related to 7c4542146f0dde962205e5a90839349631ae60fb

<del>
Add private `ActiveSupport::ProcessClock` module

In the codebase, we use `Process::CLOCK_PROCESS_CPUTIME_ID`,
but it doesn't work on Windows so we should add conditions
to support Windows. It is a good idea to handle all those cases
in one place to prevent spreading those conditions in the codebase.

Related to #34374, #34410
</del>

### CODE REVIEW

**Review by gmcgibbon — COMMENTED** [review]
Two concerns:

1. `Time.now` is a lot more readable than `Process.clock_gettime(Process::CLOCK_MONOTONIC)`. Would it be a good idea to add a new method to `Time` instead of typing this everywhere?
2. I remember in https://github.com/rails/rails/pull/34410, I had to work around something with `clock_gettime` on Windows. I think this a second good reason to add a new method to `Time` for easier fallback support.

**Comment by bogdanvlviv** [comment]
Thanks for the review!

I like the idea of preventing spreading conditions that provides support for other platforms in the codebase, but I don't think that extending of `Time` class is a good idea, I think would be better just to create a private module with methods that would work on platforms we want to support.

Please let me know what do you think about https://github.com/rails/rails/pull/34618/commits/dc424fd37beaa169d80423d3e8bbd63c61981e02?

File: activesupport/lib/active_support/process_time.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module ProcessTime # :nodoc:
+    class << self
+      def clock_monotonic
```
Comment (gmcgibbon) [inline activesupport/lib/active_support/process_time.rb]: I think we can just name this `monotonic`

File: activesupport/lib/active_support/process_time.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module ProcessTime # :nodoc:
```
Comment (gmcgibbon) [inline activesupport/lib/active_support/process_time.rb]: Does it make sense to name this `ProcessClock`?

File: activesupport/lib/active_support/process_time.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    class << self
+      def clock_monotonic
+        Process.clock_gettime(Process::CLOCK_MONOTONIC)
+      end
+
+      if defined?(Process::CLOCK_PROCESS_CPUTIME_ID) && !Gem.win_platform?
+        def clock_process_cputime_id
```
Comment (gmcgibbon) [inline activesupport/lib/active_support/process_time.rb]: This doesn't return `Process::CLOCK_PROCESS_CPUTIME_ID`, so I'd rather name this something else. Maybe `process_time`? I think [this](https://stackoverflow.com/a/7506987) breaks down the difference between constants correctly.

**Review by gmcgibbon — COMMENTED** [review]
Refactor LGTM. You're right, this doesn't belong in `Time`.

File: activesupport/lib/active_support/process_time.rb (hunk 2 above)
Comment (bogdanvlviv) [reply activesupport/lib/active_support/process_time.rb]: Yeah, seems It'll be better to name it `ProcessClock`

**Review by gmcgibbon — APPROVED** [review]

**Comment by gmcgibbon** [comment]
👍 Please squash your commits

**Comment by bogdanvlviv** [comment]
Sure, I just squashed them!

**Comment by kamipo** [comment]
> 1. `Time.now` is a lot more readable than `Process.clock_gettime(Process::CLOCK_MONOTONIC)`. Would it be a good idea to add a new method to `Time` instead of typing this everywhere?

We could also use `Concurrent.monotonic_time` which is already used in Active Record.

https://github.com/rails/rails/pull/31221/files#diff-c226a4680f86689c3c170d4bc5911e96R108

**Comment by bogdanvlviv** [comment]
> We could also use `Concurrent.monotonic_time` which is already used in Active Record.

Since we use `Concurrent.monotonic_time`, it looks like a better idea. Thanks for the tip.

File: activesupport/lib/active_support/notifications/instrumenter.rb [hunk 4]
```diff
@@ -137,17 +132,11 @@ def parent_of?(event)
...
-        if clock_gettime_supported?
-          def now_cpu
-            Process.clock_gettime(Process::CLOCK_PROCESS_CPUTIME_ID)
-          end
-        else
-          def now_cpu
-            0
-          end
+        def now_cpu
+          ActiveSupport::ProcessClock.process_time
```
Comment (kaspth) [inline activesupport/lib/active_support/notifications/instrumenter.rb]: Don’t think adding a top level file/constant is needed. Let’s just revert this change.

File: activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb [hunk 5]
```diff
@@ -166,9 +166,9 @@ def clear_cache!
 
       def explain(arel, binds = [])
         sql     = "EXPLAIN #{to_sql(arel, binds)}"
-        start   = Time.now
+        start   = Concurrent.monotonic_time
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb]: Can we use a CPU clock here given we may be waiting for IO?

Comment (rafaelfranca) [reply activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb]: Yes, you can.

File: activesupport/lib/active_support/notifications/instrumenter.rb (hunk 4 above)
Comment (bogdanvlviv) [reply activesupport/lib/active_support/notifications/instrumenter.rb]: You are right, those changes are unnecessary. It might make sense only if we have at least two cases where we use `Process.clock_gettime(Process::CLOCK_PROCESS_CPUTIME_ID)`.
